src/prototype_rust_sdk/workflow_future.rs

```diff
@@ -0,0 +1,244 @@
+                                .expect("Cancel rx not dropped");
+                        }
+                        Variant::SignalWorkflow(_) => {}
+                        Variant::RemoveFromCache(_) => {
+                            // Will make more sense once we have completions for evictions
+                            self.outgoing_completions
+                                .send(WfActivationCompletion::from_cmds(vec![], run_id))
+                                .expect("Completion channel intact");
+                            return Ok(WfExitValue::Evicted).into();
+                        }
+                    }
+                } else {
+                    return Err(anyhow!("Empty activation job variant")).into();
+                }
+            }
+
+            // TODO: Trap panics in wf code here?
+            let mut res = self.inner.poll_unpin(cx);
+
+            let mut activation_cmds = vec![];
+            while let Ok(cmd) = self.incoming_commands.try_recv() {
+                match cmd {
+                    RustWfCmd::CancelTimer(tid) => {
+                        activation_cmds.push(workflow_command::Variant::CancelTimer(CancelTimer {
+                            timer_id: tid.clone(),
+                        }));
+                        self.unblock(UnblockEvent::Timer(tid));
+                        // Re-poll wf future since a timer is now unblocked
+                        res = self.inner.poll_unpin(cx);
+                    }
+                    RustWfCmd::CancelActivity(aid) => {
+                        activation_cmds.push(workflow_command::Variant::RequestCancelActivity(
+                            RequestCancelActivity {
+                                activity_id: aid.clone(),
+                            },
+                        ));
+                    }
+                    RustWfCmd::NewCmd(cmd) => {
+                        activation_cmds.push(cmd.cmd.clone());
+
+                        let command_id = match cmd.cmd {
+                            workflow_command::Variant::StartTimer(StartTimer {
+                                timer_id, ..
+                            }) => CommandID::Timer(timer_id),
+                            workflow_command::Variant::ScheduleActivity(ScheduleActivity {
+                                activity_id,
+                                ..
+                            }) => CommandID::Activity(activity_id),
+                            _ => unimplemented!("Command type not implemented"),
+                        };
+                        self.command_status.insert(
+                            command_id,
+                            WFCommandFutInfo {
+                                unblocker: cmd.unblocker,
+                            },
+                        );
+                    }
+                    RustWfCmd::ForceTimeout(dur) => {
+                        // This is nasty
+                        std::thread::sleep(dur);
```

Comment:
Won't this block everything else in the test?

Reply:
Only inside the test it runs in. Each test gets its own thread. Blocking everything is the point in this case.